### ISSUE TITLE
atanh(- 1.0110463357335220588733500035655b-33+1b0*%i) hangs

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2719]
Imported from SourceForge on 2024-07-06 18:16:00
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2024-05-05 20:43:52
Original: https://sourceforge.net/p/maxima/bugs/4298

---

With `fpprec` set to 32, it seems as if `atanh(-
1.0110463357335220588733500035655b-33+1b0*%i)` hangs maxima.  Not sure
why.  A back trace shows:
```
 11: (EQUAL #<unavailable-arg> #<unavailable-arg> #<unavailable-arg>)[:EXTERNAL]
 12: (FPQUOTIENT (324518553658426726783156020576256 -110545049) (622037016090019667080624692264960 20))
 13: (FPATAN (-324518553658426726783156020576256 -109))
 14: (FPATAN (-324518553658426726783156020576256 111))
 15: (FPATAN2 (324518553658426726783156020576256 2) (-649037107316853453566312041152512 -109))
 16: (COMPLEX-ATANH ((BIGFLOAT SIMP 109) -425902426396778669432926343206304 -109) ((BIGFLOAT SIMP 109) 324518553658426726783156020576256 1))
 17: (BIG-FLOAT-ATANH ((BIGFLOAT SIMP 109) -425902426396778669432926343206304 -109) ((BIGFLOAT SIMP 109) 324518553658426726783156020576256 1))
 18: (BIG-FLOAT-EVAL %ATANH ((MPLUS SIMP) ((BIGFLOAT SIMP 109) -425902426396778669432926343206304 -109) ((MTIMES SIMP) ((BIGFLOAT SIMP 109) 324518553658426726783156020576256 1) $%I)))
 19: (SIMP-%ATANH ((%ATANH) ((MPLUS SIMP) ((BIGFLOAT SIMP 109) -425902426396778669432926343206304 -109) ((MTIMES SIMP) ((BIGFLOAT SIMP 109) 324518553658426726783156020576256 1) $%I))) #<unused-arg> NIL)
```
It looks like we're doing a quotient of something like 2^-110545049 by
2^20.  That's are really, really tiny number.  A taylor expansion of
`atanh(x+%i)` says the answer should be approximately `%pi*%i/4+x/2`.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 18:16:02
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2024-05-05 21:27:19
Original: https://sourceforge.net/p/maxima/bugs/4298/#2bf0

---

Maxima is getting stuck in `fpatan` with an argument of `-7.7037197775489434122239117703397b-34`.  It's using the Taylor series and it never converges because the old sum never equals the new sum (after adding a new term).

The Taylor series is the alternating series, so we should just stop if the term is much smaller than the current sum.  This comparison requires a multiply so that would slow things down, but that's better than never finishing.

Since we know that `|x| <= 1/2`, the n'th term of the series is strictly less than `(1/2)^(2*n+1)` we can compute a value of n where the term is less than `eps * x` where `eps` is, roughly, `2^-fpprec`

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 18:16:08
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2024-05-06 14:31:22
Original: https://sourceforge.net/p/maxima/bugs/4298/#d2b3

---

Fixed in commit [ef7803c516ec9f8e7256354e0c48bc1ded3ab7c2] by terminating the loop if we computed enough terms.
